### DISCUSSION TITLE
Syncing a fork leaves me one commit ahead of upstream/master

### DISCUSSION THREAD

**Discussion by plestedr** [community/community#22440] [New to GitHub]
<p>I have followed the instructions in <a href="https://help.github.com/articles/syncing-a-fork/" rel="nofollow noopener">https://help.github.com/articles/syncing-a-fork/</a> however my repository ends up being one commit ahead of the upstream/master each time I do it. If I resync with a quickly changing master I get a bunch of commits that only pertain to the resyncs. Then when I do want to create a pull request I get all of those commits in the pull. For an example, check out <a href="https://github.com/plestedr/Ada_Drivers_Library" rel="nofollow noopener">https://github.com/plestedr/Ada_Drivers_Library</a> currently identical to the master but still 8 commits ahead.</p>
<p>The only way I’ve found to work around this is to delete my repository and fork it again off of the upstream/master. That works but it seems like overkill. I tried rebasing on upstream/master, but that’s not working either. </p>
<p>What am I missing? The online docs don’t address this topic so I think its something on my end. I get the same results every time.</p>
<p>Hope y’all can point me to somewhere in the docs that can help.</p>
<p>Bob</p>

**Comment by lee-dohm** [comment]
<p>For whatever reason, you are creating merge commits when you merge the upstream version of <code>master</code> into your local copy. Once you create the first merge commit, all future sync merges will be prevented from fast-forwarding because you have a commit that is unique to your fork that does not exist on the upstream repository.</p>
<p>What you can do, if you’re certain you want your fork to exactly contain what is in the upstream repository is issue:</p>
<pre><code>git checkout master
git reset --hard upstream/master
git push --force
</code></pre>
<p>This will:</p>
<ol>
<li>Change your current branch to <code>master</code>
</li>
<li>Move <code>master</code> to point to exactly the same commit as <code>upstream/master</code>
</li>
<li>Force push that change to your GitHub fork so that your local <code>master</code>, <code>upstream/master</code>, and <code>origin/master</code> all point to the exact same commit</li>
</ol>
<p>:rotating_light: <strong>WARNING</strong> :rotating_light: This can potentially lose any commits that you have on your fork with changes unique to your fork that do not exist in the upstream repository. (But it sounds like that is what you want.)</p>
<p>I hope that helps!</p>

**Comment by bsushmith** [comment, marked as answer]
<p>Thanks for the answer. This helped me resolve the issue the first time it occurred for me.</p>
<p>I recently pushed some changes to an upstream repo and the changes got merged into that repo. But after it was merged, my fork was showing ‘1 commit ahead of [upstream], 1 commit behind [upstream]’, even though virtually everything was same between my fork and the upstream repo. So I followed the same solution thinking it would resolve this, and then it became ‘1 commit behind [upstream]’. It basically removed my own commits from my fork. So now, I did a fetch upstream before doing the reset again, and it shows the status correctly now.</p>
<p>So, I think it would be safe to add fetch before reset like below…</p>
<pre><code>git checkout master
git fetch upstream
git reset --hard upstream/master
git push --force
</code></pre>
<p>If there’s anything apparently incorrect, please do let me know, so that I can completely avoid this issue altogether.</p>

**Comment by lee-dohm** [comment]
<p>Yep, your instructions are probably slightly more complete. Thanks for sharing your solution with everyone!</p>

**Comment by collinalexbell** [comment]
<p>How are you merging? If you are doing it using the github interface (with a pull request into your fork from upstream) then the merge will automatically use --no-ff, so it will create an extra comit. <a href="https://ariya.io/2013/09/fast-forward-git-merge" rel="nofollow noopener">Source</a></p>
<p>You should, on your machine, add a remote that points to upstream, then you should pull from that remote. This will create a fast-forward commit if you haven’t changed the master branch at all.</p>
<p>Make sure that all commits you make to your fork are on new branches. If you make a commit to a branch that is actively developed on upstream (say master for example) and then upstream/master gets changed before your pull request gets merged into upstream, then your master and the upstream/master will have different histories.</p>

**Comment by wei** [comment]
<p>My Github App solves your problem : )</p>
<p><a href="https://github.com/wei/pull" rel="nofollow noopener">https://github.com/wei/pull</a></p>
<p>:robot: a GitHub App built with probot that keeps your repository up-to-date with upstream changes via automated PRs.</p>

**Comment by dhuang612** [comment]
<p>Hi, I know this is an older post but I was hoping to get some assistance. I followed your updates instructions however the final step git push --force gives me the error that I don’t have access to the repo?</p>

**Reply by eddiestefanescu** [reply]
I know it's a few years later, but you probably want to make sure that you pull from upstream and push to origin (not back to upstream where you don't have access).

**Comment by bassare** [comment]
<p>git checkout master<br>
git fetch upstream<br>
git reset --hard upstream/master<br>
git push --force</p>

**Comment by ivelin** [comment]
<p>Thanks for the solution. Works!</p>
<p>Since github changed the default <code>master</code> branch name to <code>main</code>, the solution for new repos looks like this:</p>
<p>git checkout main<br>
git fetch upstream<br>
git reset --hard upstream/main<br>
git push --force</p>

**Comment by xexyl** [comment]
I know this is an old issue but I had the same problem too .. have in the past had to delete my fork and recreate and then clone again and then set up for GitHub. This was the solution. I should have looked into this earlier on.

But I wanted to add to that if you get an error about `upstream` you can add it like:

```sh
git remote add upstream <repo_url>
```

and then proceed to the above four commands, using either `master` or `main` (or whatever the repo's branch is).

Of course it might be worth pointing out that this will also wipe out changes that have not been committed so as usual have a backup (tar it up or copy the directory and then copy the files back over).

Thanks!

# UPDATE

Fixed order of command line thanks to @arnaudrichard. It's what I get for doing it too early.

**Reply by arnaudrichard** [reply]
`git remote add upstream <repo_url>`

**Reply by xexyl** [reply]
> `git remote add upstream <repo_url>`

Hahaha thanks mate. You're absolutely right. To not confuse others (if they don't see the reply) I'll fix it. 

It was almost certainly stupid o'clock here .. like it is now.
